File: ui/src/lib/ipc/floor.ts
import { Command } from '../constants/commands'
import { call } from './transport'
import type { FloorView, RoomIconView, RoomKindView } from './types'

// The grid you painted, ranked by the game's own documented rules. Pure on the Rust side: it
// reads nothing, so calling it again with the same grid answers the same thing.
export const floorCandidates = (
  cells: (RoomKindView | null)[],
): Promise<FloorView> => call(Command.FloorCandidates, { cells })

// The game's own minimap icons, one row per room kind. Asked once when the screen opens and
// never again: the pictures do not change with what is painted, and the answer is empty on a
// machine without the game — which is not a failure, it is the screen drawing its own symbols.
export const roomIcons = (): Promise<RoomIconView[]> => call(Command.RoomIcons)

File: crates/app/src/commands/floor.rs
//! The painted grid's ranking, and the pictures the grid draws with.
//!
//! Wiring only: the grid arrives from the screen, the rules are embedded in `floor`, and the
//! icons are names the pure crate decides. Nothing here judges anything.

use ipc::{FloorView, IpcError, RoomIconView, RoomKindView};

use crate::icons::icon_url;
use crate::state::ResourcesState;

#[tauri::command]
pub(crate) fn floor_candidates(cells: Vec<Option<RoomKindView>>) -> Result<FloorView, IpcError> {
    Ok(ipc::floor_view(cells))
}

/// The fourteen room kinds with the game's own minimap icon for each.
///
/// Its own command rather than a field on `FloorView`: the icons do not depend on what is
/// painted, and `floor_candidates` runs again for every cell a dragged pointer crosses.
///
/// **No game, no URL.** The screen falls back to its own symbols either way, but a link that
/// is certain to 404 makes it fall back a frame *later* — every cell drawn once empty and then
/// again with the symbol, on a machine that never had the picture to begin with.
#[tauri::command]
pub(crate) fn room_icons(
    app: tauri::AppHandle,
    resources: tauri::State<'_, ResourcesState>,
) -> Result<Vec<RoomIconView>, IpcError> {
    let installed = resources.get(&app).is_some();
    Ok(ipc::room_icons(
        |r| {
            if installed {
                icon_url(r)
            } else {
                None
            }
        },
    ))
}
